### PR TITLE
Trajectory replacement with full blending at message arrival time

### PR DESCRIPTION
**Pull request by vedh1234**
## Description

Port of the legacy ROS 1 "update existing trajectory" behavior to the ROS 2 `joint_trajectory_controller`.

Related to #84 

### What this PR does:
- Enables the `allow_trajectory_replacement` parameter (default: `true`).
- **Merge-at-Arrival Architecture**: When a new trajectory arrives, it is immediately spliced with the currently active trajectory rather than deferring/parking the goal. 
- Constructs a continuous, single trajectory for the controller to execute:
  - **Prefix**: The arm continues exactly along its old path until the new trajectory's start time.
  - **Bridge**: A smooth, velocity-continuous transition anchor is generated exactly at the handoff point.
  - **Suffix**: Joints omitted from the new trajectory continue and complete their original path.
- Eliminates the legacy ROS 2 behavior where a future-stamped trajectory would immediately kill the active trajectory and create a slow spline ramp to the first point.
- Immediate trajectories (stamp = 0 or stamp <= now) behave exactly as before.

### ROS 1 Feature Parity Achieved in this PR:
- **Smooth Future-Stamped Blending**: Commanded joints faithfully follow their old path until the new trajectory's exact start time.
- **Partial Goal Continuity**: Omitted joints do not freeze; they outlast the new trajectory and complete their original motion (Suffix).
- **Speed Scaling Safety**: The blending anchor uses the trajectory's internal scaled cursor (`traj_time_`) rather than wall-clock time, guaranteeing **zero position jumps** even under speed scaling (an improvement over some legacy implementations).

### What this PR does NOT target (ROS 1 Parity Gaps):
- **Omitted Joint Spline Fidelity**: In ROS 1, omitted joints perfectly retained their original independent trajectory segments. Due to ROS 2's monolithic trajectory architecture, omitted joints must be re-sampled onto the new trajectory's time grid. If the new trajectory has sparse waypoints, the omitted joint will re-interpolate between them, which may result in minor deviations from its original intermediate path.

### Is this user-facing behaviour change?
Yes. When `allow_trajectory_replacement` is true (default), publishing a future-stamped trajectory will now splice into the current motion.

### Did you use Generative AI?
Yes, Google Gemini and Claude were used to make improvements in manually written code, architectural analysis, and edge-case review. 

### Additional Information
- The disabled test `test_execute_partial_traj_in_future` referenced in #84 has been enabled and passes.
- 3 new parameterized integration tests (`blend_no_position_jump_under_speed_scaling`, `blend_omitted_joint_outlasts_shorter_new_trajectory`, `blend_commanded_joint_follows_old_path`) were added to  validate the prefix/suffix math.

## TODOs
- [x] Fork the repository.
- [x] Modify the source; please focus on the specific change you are contributing.
- [x] Ensure local tests pass. (`colcon test` and `pre-commit run`)
- [x] Commit to your fork using clear commit messages.
- [x] Send a pull request, answering any default questions in the pull request interface.
- [x] Pay attention to any automated CI failures reported in the pull request, and stay involved in the conversation.

### CODE REVIEW

**Review by saikishor — DISMISSED** [review]
LGTM

**Comment by saikishor** [comment]
@christophfroehlich what do you think?

File: joint_trajectory_controller/doc/trajectory.rst [hunk 1]
```diff
@@ -116,19 +116,17 @@ Trajectory Replacement
...
-
-  + The controller does not splice or stitch the current and new trajectories together. At the handoff time, the new trajectory fully replaces the old one.
-  + Omitted joints in a partial goal are filled with their current position at the handoff time (hold-in-place), rather than continuing to follow the old trajectory.
-  + Only a single deferred trajectory is stored. A newer deferred trajectory overwrites any previously deferred one.
-  + When a new action goal is accepted, the currently active goal is immediately aborted instead of continuing to emit feedback until the handoff time.
-  + While a deferred action goal is waiting for its start time, it does not publish action feedback and is not evaluated for path tolerances.
-  + The handoff trigger uses the controller's clock time. If the controller is running with a ``speed_scaling`` other than 1.0, the handoff may occur out of sync with the active trajectory's execution state.
+  This behavior is implemented in ROS 2 via the ``allow_trajectory_replacement`` parameter
+  (default ``true``; see `#84 <https://github.com/ros-controls/ros2_controllers/issues/84>`__).
```
Comment (christophfroehlich) [inline joint_trajectory_controller/doc/trajectory.rst]: will we close this issue now? than we should remove it from here.

File: joint_trajectory_controller/test/test_trajectory_controller.cpp [hunk 2]
```diff
@@ -2101,6 +2101,191 @@ TEST_P(TrajectoryControllerTestParameterized, test_execute_partial_traj_in_futur
...
+
+  std::vector<std::vector<double>> new_traj{{{0., 0., 0.}, {-5., -5., -5.}}};
+  publish(rclcpp::Duration::from_seconds(1.0), new_traj, rclcpp::Time());
+  traj_controller_->wait_for_trajectory(executor);
+  t = updateControllerAsync(rclcpp::Duration::from_seconds(0.03), t);
+  const auto ref_after = traj_controller_->get_state_reference();
+
+  for (size_t i = 0; i < ref_before.positions.size(); ++i)
+  {
+    EXPECT_NEAR(ref_before.positions[i], ref_after.positions[i], 2.0)
```
Comment (christophfroehlich) [inline joint_trajectory_controller/test/test_trajectory_controller.cpp]: can this expectations changed to a more exact one?

**Review by christophfroehlich — CHANGES_REQUESTED** [review]
In general LGTM. Two minor comments, and imho the copilot has some valid suggestions too.

File: joint_trajectory_controller/doc/trajectory.rst (hunk 1 above)
Comment (vedh1234) [reply joint_trajectory_controller/doc/trajectory.rst]: Removed as this PR closes #84

File: joint_trajectory_controller/test/test_trajectory_controller.cpp (hunk 2 above)
Comment (vedh1234) [reply joint_trajectory_controller/test/test_trajectory_controller.cpp]: Okay I have updated.

File: joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml [hunk 3]
```diff
@@ -77,6 +77,15 @@ joint_trajectory_controller:
...
     description: "Allow joint goals defining trajectory for only some joints.",
   }
+  allow_trajectory_replacement: {
+    type: bool,
+    default_value: true,
+    read_only: true,
+    description: "When a new trajectory arrives while one is still executing, blend with the old one
+      instead of hard-replacing it: joints not commanded by the new trajectory keep following the old
+      trajectory (instead of holding at the current position), and the transition uses the existing
+      bridge so it stays smooth even under speed scaling. Opt-in; default keeps the legacy behavior.",
```
[the remark this answers is not in the document]
Comment (vedh1234) [reply joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml]: I have updated in the docs

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 4]
```diff
@@ -1647,6 +1600,168 @@ void JointTrajectoryController::fill_partial_goal(
...
+      {
+        auto p = op;
+        for (size_t j = 0; j < dof_; ++j)
+        {
+          if (blend_commanded_[j])
+          {
+            if (!p.positions.empty()) p.positions[j] = new_last.positions[j];
+            if (!p.velocities.empty()) p.velocities[j] = 0.0;
+            if (!p.accelerations.empty()) p.accelerations[j] = 0.0;
+          }
```
[the remark this answers is not in the document]
Comment (vedh1234) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:1781]: Instead of above `blend_with_active_trajectory` now returns false when a goal carries no positions and joints are omitted, and the caller falls back to the legacy path.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 5]
```diff
@@ -267,60 +267,25 @@ controller_interface::return_type JointTrajectoryController::update(
...
-
-  // The trajectory message to be installed into current_trajectory_ this cycle (if any).
-  std::shared_ptr<trajectory_msgs::msg::JointTrajectory> traj_msg_to_install = nullptr;
-
   // Discard, if a goal is pending but still not active (somewhere stuck in goal_handle_timer_)
   if (
-    current_trajectory_msg != *new_external_msg && *new_external_msg != pending_traj_msg_ &&
-    (rt_has_pending_goal_ && !active_goal) == false)
+    current_trajectory_msg != *new_external_msg && (rt_has_pending_goal_ && !active_goal) == false)
   {
```
[the remark this answers is not in the document]
Comment (vedh1234) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:283]: This is unchanged from master  also not an issue. The buffer is only ever null between `on_activate()'s` `writeFromNonRT(nullptr)` and the first real message and at that point `current_trajectory_->get_trajectory_msg()` is also null, so the != is false and the block is skipped. Every other writer (`add_new_trajectory_msg`, and the `reset()`+ `initRT()` pairs in `update()`) always writes non-null.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 6]
```diff
@@ -267,60 +267,25 @@ controller_interface::return_type JointTrajectoryController::update(
...
-    if (is_internal_hold(*new_external_msg))
-    {
-      // Internal hold/success/decelerate: install but never cancel a deferred trajectory.
-      traj_msg_to_install = *new_external_msg;
-    }
-    else if (
+    if (
       params_.allow_trajectory_replacement && has_active_trajectory() &&
-      rclcpp::Time((*new_external_msg)->header.stamp, time.get_clock_type()) > time)
+      current_trajectory_->has_nontrivial_msg() && !rt_is_holding_)
```
[the remark this answers is not in the document]
Comment (vedh1234) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:287]: Not an issue. For immediate trajectories (stamp=0 or <=now), new_start = cursor, making the prefix empty and the suffix empty (for full-joint goals), so the blend degenerates to the same result with an optional velocity-continuous bridge point. Negative time_from_start is not possible as the two rebasing steps cancel to new_start - cursor + original_tfs, and new_start >= cursor always holds. The behavioral change for partial goals (omitted joints follow old trajectory vs freeze) is intentional.

**Comment by vedh1234** [comment]
While resolving comments I found some improvements and a bug. I have fixed it in the recent update: 

1. Improvement: Reversed the interpolation loops to advance monotonically point-by-point. 
2. Bridge & Edge Cases: fixed the speed-scaling math for future-stamped trajectories and added a safe fallback to hardware states to prevent crashes on velocity-only goals. Added a check on `last_sample_idx_` to prevent off-cursor probes from corrupting the RT thread's search index (pre-existing bug just was discovered by new function calls).
3. Tests: Relaxed the EXPECT_NEAR peak tolerances. Added the test `blend_partial_goal_outlasting_old_trajectory` which actually caught the `last_sample_idx_` issue.

**Review by christophfroehlich — DISMISSED** [review]
It seems that some flakiness was introduced, I never saw that failing before:

```
     [  FAILED  ] ScaledJTCTests/TestTrajectoryActionsTestScalingFactor.test_scaling_execution_time_succeeds/0, where GetParam() = 0.25
    
     1 FAILED TEST
    -- run_test.py: return code 1
    -- run_test.py: inject classname prefix into gtest result file '/home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/build/joint_trajectory_controller/test_results/joint_trajectory_controller/test_trajectory_actions.gtest.xml'
    -- run_test.py: verify result file '/home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/build/joint_trajectory_controller/test_results/joint_trajectory_controller/test_trajectory_actions.gtest.xml'
  >>>
build/joint_trajectory_controller/test_results/joint_trajectory_controller/test_trajectory_actions.gtest.xml: 143 tests, 0 errors, 1 failure, 0 skipped
- joint_trajectory_controller.ScaledJTCTests/TestTrajectoryActionsTestScalingFactor test_scaling_execution_time_succeeds/0 (/home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/src/ros2_controllers/joint_trajectory_controller/test/test_trajectory_actions.cpp:1343)
  <<< failure message
    /home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/src/ros2_controllers/joint_trajectory_controller/test/test_trajectory_actions.cpp:1376
    Expected equality of these values:
      static_cast<int32_t>(expected_index)
        Which is: 0
      feedback_msg->index
        Which is: 1
  >>>

```

**Comment by saikishor** [comment]
> It seems that some flakiness was introduced, I never saw that failing before:
> 
> ```
>      [  FAILED  ] ScaledJTCTests/TestTrajectoryActionsTestScalingFactor.test_scaling_execution_time_succeeds/0, where GetParam() = 0.25
>     
>      1 FAILED TEST
>     -- run_test.py: return code 1
>     -- run_test.py: inject classname prefix into gtest result file '/home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/build/joint_trajectory_controller/test_results/joint_trajectory_controller/test_trajectory_actions.gtest.xml'
>     -- run_test.py: verify result file '/home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/build/joint_trajectory_controller/test_results/joint_trajectory_controller/test_trajectory_actions.gtest.xml'
>   >>>
> build/joint_trajectory_controller/test_results/joint_trajectory_controller/test_trajectory_actions.gtest.xml: 143 tests, 0 errors, 1 failure, 0 skipped
> - joint_trajectory_controller.ScaledJTCTests/TestTrajectoryActionsTestScalingFactor test_scaling_execution_time_succeeds/0 (/home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/src/ros2_controllers/joint_trajectory_controller/test/test_trajectory_actions.cpp:1343)
>   <<< failure message
>     /home/runner/work/ros2_controllers/ros2_controllers/.work/target_ws/src/ros2_controllers/joint_trajectory_controller/test/test_trajectory_actions.cpp:1376
>     Expected equality of these values:
>       static_cast<int32_t>(expected_index)
>         Which is: 0
>       feedback_msg->index
>         Which is: 1
>   >>>
> 
> ```



Ohhh nooooooo

**Comment by vedh1234** [comment]
I am looking into it, was trying to reproduce it both ways with and without my change, in both cases it's passing. Still I will try different scenarios to see if it fails.

**Comment by vedh1234** [comment]
I ran the scaling soak tests 200 times each and counted the `feedback_msg->index` failures:

  - upstream/master with none of the blend code compiled in: 2/200 ( 2 times failed) 
  - this branch with the extra fix below: 2/200

So to further check I added some temporary ns logging in the feedback callback to see why above failures and managed to catch one:

  `desired.time_from_start` = 0.099999107     `index` = 1

That can't happen in a single update cycle. At 0.099999107 on the [0.1, 0.2] (time) trajectory, sample() goes into the `sample_time` < `first_point_timestamp` branch so `next_point_index` is 0. You only get index = 1 once rel >= 0.1. So the two fields are coming from different cycles. So this points towards a torn read. 

I looked at what the PR #2160 added which was recently merged:

`auto feedback = std::make_shared<FollowJTrajAction::Feedback>();`   changed to:
`const auto & feedback = active_goal->preallocated_feedback_;`

Earlier every cycle gave the publisher its own message, so nothing else could touch it. Now it's one shared message, and the RT thread keeps writing into it while the 20 Hz timer is serializing. setFeedback() only locks the pointer assignment, not the field writes, and desired comes before index in the message.

One caveat though, I only got exact timing on 1 of the failures, the rest I just counted. There's another possible cause with the same signature: the test uses `lower_bound` which puts an exact waypoint time in the previous segment, while sample() uses >= t0 and puts it in the next. So this can also be one of the reason. 

So I am not yet sure either the torn read is the issue or this `lower_bound` in test but the soak tests are pointing that this existed before the PR.  

Separate thing, I found a missing check which now I added just to avoid any index shift in this PR while digging, and as well I have fixed the copilots suggestion. Also Added `blend_skipped_for_immediate_full_goal` which fails without the fix.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 7]
```diff
@@ -1658,6 +1609,197 @@ void JointTrajectoryController::fill_partial_goal(
...
+  // sample the bridge anchor before the sweep below advances the trajectory's search cursor
+  TrajectoryPointConstIter start_segment_itr, end_segment_itr;
+  const bool have_bridge = current_trajectory_->sample(
+    new_start, interpolation_method_, blend_bridge_, start_segment_itr, end_segment_itr, false);
+
+  fill_omitted_joints_from_old(trajectory_msg, new_start);
+  sort_to_local_joint_order(trajectory_msg);
+  const auto new_last = trajectory_msg->points.back();
+  const rclcpp::Time new_end =
+    new_start + rclcpp::Duration(trajectory_msg->points.back().time_from_start);
```
[the remark this answers is not in the document]
Comment (vedh1234) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:1734]: I have fixed this issue in the latest update.

File: doc/release_notes.rst [hunk 8]
```diff
@@ -52,11 +52,13 @@ joint_trajectory_controller
...
-  continues executing until the start time arrives, then the new trajectory is executed.
-  This is a partial port of the ROS 1 trajectory replacement behavior.
-  (`#2401 <https://github.com/ros-controls/ros2_controllers/pull/2401>`_)
+* Fixed the final segment of every trajectory being cut short: the next-cycle lookahead sample advanced the shared segment search cursor to the end, so the following cycle jumped the reference to the last waypoint and succeeded the goal early. Independent of ``allow_trajectory_replacement``. (`#2419 <https://github.com/ros-controls/ros2_controllers/pull/2419>`_)
+* Ported the ROS 1 trajectory-replacement behavior via the ``allow_trajectory_replacement``
+  parameter. A trajectory arriving while another is executing is spliced into
+  the active one instead of discarding it: the old path is followed up to the new start time, a
+  velocity-continuous bridge is sampled at the handoff, and joints omitted from a partial goal
+  continue and finish their original motion. (`#2419
+  <https://github.com/ros-controls/ros2_controllers/pull/2419>`_)
```
Comment (bmagyar) [inline doc/release_notes.rst]:
```suggestion
* Ported the ROS 1 trajectory-replacement behavior via the ``allow_trajectory_replacement``  parameter. A trajectory arriving while another is executing is spliced into the active one instead of discarding it: the old path is followed up to the new start time, a velocity-continuous bridge is sampled at the handoff, and joints omitted from a partial goal continue and finish their original motion. (`#2419 <https://github.com/ros-controls/ros2_controllers/pull/2419>`_)
```

**Review by bmagyar — APPROVED** [review]
thank you!
